src/clj_gatling/simulation.clj

```diff
@@ -123,62 +137,90 @@
    user-id]
-  (let [c (async/chan)
-        should-run-now? (if concurrency-distribution
-                          (let [modifier-fn (if (= 2 (arg-count concurrency-distribution))
-                                              (fn [{:keys [progress context]}]
-                                                (concurrency-distribution progress context))
-                                              concurrency-distribution)]
-                            #(let [[progress duration] (runners/calculate-progress runner
-                                                                                   @sent-requests
-                                                                                   simulation-start
-                                                                                   (LocalDateTime/now))
-                                   target-concurrency (* concurrency (modifier-fn {:progress progress
-                                                                                   :duration duration
-                                                                                   :context context}))]
-                               (> target-concurrency @concurrent-scenarios)))
-                          (constantly true))]
-    (go-loop []
-      (if (should-run-now?)
-        (do
-          (swap! concurrent-scenarios inc)
-          (let [result (<! (run-scenario-once options scenario user-id))]
+  (let [c                (async/chan)
+        continue?        (continue-fn {:runner runner
+                                       :simulation-start simulation-start
+                                       :force-stop force-stop})
+        should-run-now?  (if concurrency-distribution
+                           (let [modifier-fn (if (= 2 (arg-count concurrency-distribution))
+                                               (fn [{:keys [progress context]}]
+                                                 (concurrency-distribution progress context))
+                                               concurrency-distribution)]
+                             #(let [[progress duration] (runners/calculate-progress runner
+                                                                                    (:sent @simulation-requests)
+                                                                                    @simulation-start
+                                                                                    (LocalDateTime/now))
+                                    target-concurrency (* concurrency (modifier-fn {:progress progress
+                                                                                    :duration duration
+                                                                                    :context context}))]
+                                (atomic-concurrency-check target-concurrency concurrent-scenarios)))
+                           #(atomic-concurrency-check concurrency concurrent-scenarios))
```

Comment:
This is no longer just `(constantly true)`, even if there is no ramping function, because we purposefully create more runners than the target concurrency.